### ISSUE TITLE
Curve-Line Intersection

### ISSUE THREAD

**Issue by cbrunschen** [Pomax/BezierInfo-2#383]
The current text fairly simply reduces this problem to

> first we translate/rotate both the line and curve together, in such a way that the line coincides with the x-axis [...] the problem of finding intersections between a curve and a line has now become the problem of performing root finding on our translated/rotated curve, as we already covered in the section on finding extremities.

However, https://www.particleincell.com/2013/cubic-line-intersection/ gives a simple solution that does not require translations or rotations, but directly constructs a polynomial, by starting from the polynomial form of the curve, `p(t) = b3 t^3 + b2 t^2 + b1 t + b0` <=> `x(t) = b3.x t^3 + b2.x t^2 + b1.x t * b0.x` and analogously for `y(t)`.

The line's equation is of course `a x + b y + c == 0`. Put `x(t)` and `y(t)` into the equation for the line and expand x(t) and y(t) and then gather all the `t^1`, `t^2`, `t` and constant terms together and you get

```
   t^3 * (a * b3.x + b * b3.y) 
 + t^2 * (a * b2.x + b * b2.y) 
 +   t * (a * b1.x + b * b1.y) 
 +       (a * b0.x + b * b0.y) + c 
 == 0
```

This polynomial that can be fed into Cardano's algorithm and produce the `t` values where the line and the curve intersect.

The same can of course be done for a Quadratic curve.

**Comment by Pomax** [comment]
Right, but we're also programming, and while that solution works, it's also more expensive to run. It might be worth adding as solution, but it's basically a worse solution for computers (even if on paper it's more elegant).

**Comment by cbrunschen** [comment]
Would you care to explain what leads you to conclude that this is more expensive to run, and a worse solution for computers? Because to me, it doesn't immediately seem that way.

I don't think the sketched approach necessarily has to be particularly a lot slower / more expensive to run, plus I think this may have some accuracy advantages. 

Both approaches require finding the roots of cubic function, so that is going to take the same amount of time; any difference in expense is going to be in the calculation of the coefficients to feed into the root-finding function.

The current align-to-the-x-axis approach requires one atan2() call, as well as a sin() and and cos(), in order to create the affine transformation that is then applied to each of the points to align the curve to the x axis. The sketched approach uses additions/subtractions and multiplications – possibly a few more of them than in the current approach, but it seems like that expense may be balanced by the lack of calls to (usually fairly expensive) trigonometric functions. 

This would really have to be benchmarked, and I'm guessing this might differ depending on the choice of language / runtime environment. 

I actually wrote and ran a rudimentary benchmark program in C, and the sketched approach was consistently faster (by a factor of about 1.7, taking about 60% of the time) compared to the current approach on my laptop (a 2013 Macbook Pro running Mac OS X 10.13, using LLVM 10.0 – I know, not exactly state of the art, but it is what I use at the moment). This certainly appears to give at least one counterexample to the claim that the sketched approach is more expensive to run, though I am by no means going to conclude that it is always going to be faster – just that it doesn't seem to necessarily always be slower.

The sketched approach may also have an accuracy advantage. While any floating-point calculations may generate an inexact result, if you stick with additions/subtractions and multiplications, there's a fairly wide range of inputs where the results will be exact. However the atan2(), sin() and cos() trigonometric functions will pretty much always produce something that is an approximation – if for no other reason that there's a need to roundtrip the angles through some multiples of PI, which of course cannot be represented exactly. 

Consider for example any curves and lines with integer coordinates within +- 2^24 of the origin, where the line is neither horizontal nor vertical. 

In the current approach, both the angle, its sin() and cos() are going to be non-integer, non-exact approximations, and thus so will the coordinates of the axis-aligned curve, creating the first small error; this error is then increased trying to find the roots of the already-slightly-erroneous cubic function. 

With the sketched approach, all the differences, sums and products are going to involve integers within the 53-bit mantissa of a double, and can thus be represented exactly, with zero error - and the first approximations happen and thus the first errors are introduced in the process of finding the roots of the cubic function.

So while both approaches will of course produce an approximation as part of finding the roots of the cubic function, having less of an error in the inputs to the root finding may produce a more accurate result for the 't' value because there are fewer steps along the calculation where an error has been introduced.

**Comment by jakub-trzebiatowski** [comment]
First, @cbrunschen you are right that the algorithm you quote is elegant and efficient to compute.

But the algorithm you provided should have the same numeric stability as the one from the book and roughly similar performance. It's nearly the same algorithm, just inlined for this specific purpose.

The book says...

> first we translate/rotate both the line and curve together, in such a way that the line coincides with the x-axis

...which means that, given a line described by its parametric equation `p = p0 + dt`, the weights of the transformed Bezier curve are...

```
B'_n = rotateToXAxis(translate(B_n, delta = -p0), v = d)
```

If we implement rotation-to-X-axis non-naively, accepting the angle represented as a vector, we'll compute (pseudocode)...

```
function rotateToXAxis(p, v) {
  let sinFi = v.y / |v|
  let cosFi = v.x / |v|
  return [
    p.x * cosFi - p.y * sinFi,
    p.x * sinFi + p.y * cosFi,
  ]
}

```

Assuming that the direction vector of the line is stored in the normalized way, division `/ |v|` can be discarded.

As you see, the `(a * b_n.x + b * b_n.y)`part in your code looks a lot like the y-component of the rotation transformation in the function above.

So yes, we can optimize the approach described in the book by inlining as much as possible, **but** the book never said that you have to compute any cosine or sine to perform the rotation.

_I'm not commenting on any JavaScript code from the interactive examples as I don't consider them a part of the book._

**Comment by cbrunschen** [comment]
That is a great observation @cubuspl42 - as you say it improves both the performance and stability of the translate-and-rotate approach.

But as a "[Primer](https://en.wikipedia.org/wiki/Primer_(textbook))" is generally an introductory text I would expect anyone reading this Primer to look at the text and the sample code that it presents, take what is written there essentially at face value and implement those things in a straightforward, even "naïve", manner. Neither the optimization you propose, nor a suggestion that the algorithm described _should_ be optimized, is ever mentioned in the Primer; so while the algorithm described in the Primer _can_ be optimized, that's not what I'd expect most readers of the Primer to do.

The entire text in the Primer on this subject is:

> Curve/line intersection is more work, but we've already seen the techniques we need to use in order to perform it: first we translate/rotate both the line and curve together, in such a way that the line coincides with the x-axis. This will position the curve in a way that makes it cross the line at points where its y-function is zero. By doing this, the problem of finding intersections between a curve and a line has now become the problem of performing root finding on our translated/rotated curve, as we already covered in the section on finding extremities.

And rotating has been previously discussed in the Primer only using sines and cosines in the usual matrix multiplication fashion; so a straightforward, naïve implementation that follows what the Primer explains would use those as well.

On your `rotateToXAxis()` function, I think it needs at the very least a different name: after all it not only _rotates_ the point $b$ around the origin by the angle of $v$, but it also _scales_ it by a factor of $|v|$, as you're discarding the division by $|v|$.

For the purpose of applying that to a curve that is translated to the origin and then finding where that curve intersects the $x$ axis this does not matter much - the $t$ values where the scaled curve intersects the $x$ axis will be the same as for the un-scaled curve; which is why for this particular purpose as you say, "division `/ |v|` can be discarded". 

But that's not an immediately obvious optimization to make.

And it rather changes what the function does, and the overall algorithm too: it no longer just _translates_ and _rotates_ the curve, but in order to improve the performance, it also _scales_ the curve in a way that happens to not adversely affect the result; but this is still something that, in the context of a Primer document in particular, would really need to be _explained_ - and the function appropriately (re)_named_.

Conversely, as it is written, the function also does more work than we need for the purpose of finding the $t$ values where $y = 0$: only the $y$ coordinate matters, the $x$ coordinate is redundant, so we don't even need to calculate the x coordinates of the rotated points that we will end up not using - allowing a further optimization.

So your optimization is not only a bit non-obvious, but also leaves the process arguably more complex ('translate and rotate the curve; for performance reasons, also scaling it but that's OK we can ignore that; also ignore the resulting $x$ coordinates; then find roots from the remaining $y$ coordinates') than the suggested approach which directly combines the equation for the line with the polynomial form of the Bézier curve.

Your optimized implementation does bring the solution to much the same level of performance and stability. In my testing [this test code](https://gist.github.com/cbrunschen/0328e45d9ec1e8cfdee56aa870b9ab66) on my machine, the optimized axis-aligned solution is repeatably just a touch (between 1% and 3%, depending on choice of compiler (clang or gcc) and optimization level) slower than the suggested approach; though using [Compiler Explorer](https://godbolt.org/z/nnPPxP1YG), I see run-to-run variation where sometimes the optimized-alignment method wins instead. 

(In my tests I am of course not calculating the redundant $x$ coordinate, even for the most naïve implementation; and in all my testing I'm passing six points - four to define the Cubic Bézier curve and two to define the line - into the intersection-calculating function, to make the comparison as fair as possible, so there's no pre-calculating the polynomial form of the Bézier curve or similar for example.)

So in the end I remain confused about Pomax's rather dismissive reply. His claim that 

>  [...] while that solution works, it's also more expensive to run [...] basically a worse solution for computers [...]. 

even with your optimizations, in my testing, simply does not match what I'm measuring. Nor does his comment 

> but we're also programming

make much sense to me: the complexity of the code is very similar, and the suggested approach does not require optimization to achieve good performance: the naïve equation-combining code is as fast as the optimized x-axis-aligning one, without requiring an explanation of how to optimize it.

Of course this is Pomax's primer and he can include in it whatever he wants. But his dismissive assessment of the suggested approach just doesn't match the what I'm seeing in complexity of explanation, code complexity or performance measurements.
